docs/design/sql/10-distributed-scan.md

```diff
@@ -0,0 +1,146 @@
+
+The `IMap` is a distributed map. It could be accessed either directly, or through a secondary index. The secondary index is 
+a distributed data structure. Every member stores part of the index that is built for the local entries of the member. 
+
+The goal of the Hazelcast Mustang engine is to pick the proper access method - direct scan or index scan, start execution of 
+the scan on members, and collect the results. In the above sections we describe how it is implemented.   
+
+## 2 Access Path Selection
+
+There are two ways to scan the `IMap` - iterate over the record store directly, or use one of the secondary indexes. During the
+planning stage the proper access method is chosen. 
+
+If there are no indexes on the table, the direct scan is chosen and no further optimization is performed.    
+
+If there are indexes on the table, we analyze the predicate stored in the table of the `TableScan` operator. The entry point
+is `IndexResolver.createIndexScans`. 
+
+First, we split the predicate into conjunctive normal form (CNF). For example, the predicate `a=1 AND b=2` is split into 
+`a=1` and `b=2`, while the predicate `a=1 OR b=2` remains unchanged.
+
+Second, for every sub-predicate we find those that could be used by some index. Assume that `col` is a simple column
+expression, and `exp` is an expression that has only constants or parameters at leaves (i.e. it doesn't refer to other columns). 
+We use the following rules:
+1. `col = exp` can be used with `SORTED` and `HASH` indexes
+1. `col [comparison] exp` can be used with `SORTED` indexes
+1. `col IS NULL/TRUE/FALSE` can be treated as equality expression (with slightly different semantics for `NULL` values)
+1. `col = exp1 OR col = exp2` can be treated as union of two equality predicates 
+
+The result of this step is a map from column to candidate expressions that could be used with indexes. For example, 
+`a=1 AND b>2 AND b<4` is returned as: 
+```
+a -> [=1] 
+b -> [>2], [<4]
+```
+
+Next, we iterate over every index, and try to bind candidates to the index based on the index columns and index type.
+General rules are:
+1. `SORTED` index may use equality and comparison conditions, while `HASH` index may use only equality conditions
+1. Only a continuous prefix of index columns can be used. E.g. for the index `(a, b, c)` and the condition `a=1 AND c=2` we 
+can use only `a=1`
+1. All expressions in the prefix except for the last one must be equality conditions. E.g. for the index `(a, b)` and the 
+condition `a>1 AND b>2`, we can use only `a>1`   
+
+The result of this step is a map from index to the filter that should be used. For example, for the expression 
+`a=1 AND b>2 AND b<4`, and indexes on `(a, c)` and `(b)`, the result would be:
+```
+index(a, c) -> [a=1, NULL]
+index(b) -> [b>2 AND b<4]
+```
+
+At last, for every proposed index, we create a `MapIndexScanPhysicalRel` operator that is added to the planner search space.
```

Comment:
It would be great to have a paragraph on costing, how and why one scan is preferred over another index or map scan.

Reply:
It also worth to mention potential performance implications of having many matching index scans, how does it affect  the planing time, possible solutions for improving it, etc.

Reply:
Add an explanation of the cost model.

Reply:
Index planning doesn't have any significant impact at the moment, because we do not have joins. We will have to reconsider how we add indexes to the search space in the future and add the appropriate design documentation.